crates/sui-types/src/nitro_attestation.rs

```diff
@@ -0,0 +1,544 @@
+use std::fmt;
+use std::time::Duration;
+use webpki::types::UnixTime;
+
+use crate::error::{SuiError, SuiResult};
+
+use ciborium::value::{Integer, Value};
+use p384::ecdsa::signature::Verifier;
+use p384::ecdsa::{Signature, VerifyingKey};
+use rustls::pki_types::CertificateDer;
+use rustls::server::WebPkiClientVerifier;
+use rustls::RootCertStore;
+use x509_parser::{certificate::X509Certificate, prelude::FromDer};
+
+#[cfg(test)]
+#[path = "unit_tests/nitro_attestation_tests.rs"]
+mod nitro_attestation_tests;
+
+const ROOT_CERTIFICATE_PEM: &[u8] = include_bytes!("./nitro_root_certificate.pem");
+const P384_PUBLIC_KEY_SIZE: usize = 97;
+
+#[derive(Debug, PartialEq, Eq)]
+pub enum NitroError {
+    /// Invalid COSE_Sign1: {0}
+    InvalidCoseSign1(String),
+    /// Invalid signature
+    InvalidSignature,
+    /// Invalid attestation document
+    InvalidAttestationDoc(String),
+    /// Invalid user data
+    InvalidUserData,
+    /// Invalid certificate: {0}
+    InvalidCertificate(String),
+    /// Invalid PCRs
+    InvalidPcrs,
+}
+
+impl fmt::Display for NitroError {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        match self {
+            NitroError::InvalidCoseSign1(msg) => write!(f, "InvalidCoseSign1: {}", msg),
+            NitroError::InvalidSignature => write!(f, "InvalidSignature"),
+            NitroError::InvalidAttestationDoc(msg) => write!(f, "InvalidAttestationDoc: {}", msg),
+            NitroError::InvalidCertificate(msg) => write!(f, "InvalidCertificate: {}", msg),
+            NitroError::InvalidPcrs => write!(f, "InvalidPcrs"),
+            NitroError::InvalidUserData => write!(f, "InvalidUserData"),
+        }
+    }
+}
+
+impl From<NitroError> for SuiError {
+    fn from(err: NitroError) -> Self {
+        SuiError::AttestationFailedToVerify(err.to_string())
+    }
+}
+
+/// Given an attestation document bytes, deserialize and verify its validity according to
+/// <https://docs.aws.amazon.com/enclaves/latest/user/verify-root.html>
+/// and check the user_data is consistent with the enclave public key.
+pub fn attestation_verify_inner(
+    attestation_bytes: &[u8],
+    enclave_vk: &[u8],
+    pcrs: &[&[u8]],
+    timestamp: u64,
+) -> SuiResult<()> {
+    // Parse attestation into a valid cose sign1 object with valid header.
+    let cose_sign1 = CoseSign1::parse_and_validate(attestation_bytes)?;
+
+    // Parse attestation document payload and verify cert against AWS root of trust.
+    let doc =
+        AttestationDocument::parse_and_validate_payload(&cose_sign1.payload, timestamp, pcrs)?;
+
+    // Extract public key from cert and signature as P384.
+    let signature = Signature::from_slice(&cose_sign1.signature).expect("Invalid signature");
+    let cert = X509Certificate::from_der(doc.certificate.as_slice())
+        .map_err(|e| NitroError::InvalidCertificate(e.to_string()))?;
+    let pk_bytes = cert.1.public_key().raw;
+    if pk_bytes.len() < P384_PUBLIC_KEY_SIZE {
```

Comment:
can this be larger?

Reply:
turns out there is a proper util function in x509_parser to get the sec1 bytes! revised to that instead.